### ISSUE TITLE
Using composer installed drush

### ISSUE THREAD

**Issue by welly** [drush-ops/drush#1474]
Hello,

Having a slight issue with installing a site-specific composer installed version of Drush. The problem is that it appears to not be able to recognise the Drupal installation.

I have a site install structure like:

```
/home/web/example
/home/web/example/build
/home/web/example/build/bin/drush
/home/web/example/build/docroot
/home/web/example/build/vendor
/home/web/example/build/composer.json
/home/web/example/build/composer.lock
```

When I run /home/web/example/build/bin/drush status, I get the following returned:

```
 PHP executable         :  /usr/bin/php
 PHP configuration      :  /etc/php.ini
 PHP OS                 :  Linux
 Drush script           :  /home/web/example/build/vendor/drush/drush/drush.php
 Drush version          :  7.0.0
 Drush temp directory   :  /tmp
 Drush configuration    :
 Drush alias files      :
```

I've tried creating an alias file in /home/web/example/build/docroot/sites/default/drushrc.php and using:

```
/home/web/example/build/bin/drush --local $@ status
```
which returns

```
 Drupal version         :  7.37
 Site URI               :  http://default
 PHP executable         :  /usr/bin/php
 PHP configuration      :  /etc/php.ini
 PHP OS                 :  Linux
 Drush script           :  /home/web/example/build/vendor/drush/drush/drush.php
 Drush version          :  7.0.0
 Drush temp directory   :  /tmp
 Drush configuration    :  /home/web/example/build/docroot/sites/default/drushrc.php
 Drush alias files      :
 Drupal root            :  /home/web/example/build/docroot
 Site path              :  sites/default
```

So basically I'm pretty stumped and have not found anything that would suggest what I'm doing is incorrect. Any help/advice would be appreciated.

Cheers

**Comment by weitzman** [comment]
Drush can't yet find Drupal root automatically. Pass --root option. You can copy the script at examples/drush to wherever you'd like - and edit as needed.

**Comment by welly** [comment]
Have given it the --root option and I'm still not having any joy:

```
-bash-4.1$ /home/web/example/build/bin/drush --root=/home/web/example/build/docroot/ status
 Drupal version         :  7.38
 Site URI               :  http://default
 PHP executable         :  /usr/bin/php
 PHP configuration      :  /etc/php.ini
 PHP OS                 :  Linux
 Drush script           :  /home/web/example/build/vendor/drush/drush/drush.php
 Drush version          :  7.0.0
 Drush temp directory   :  /tmp
 Drush configuration    :
 Drush alias files      :
 Drupal root            :  /home/web/example/build/docroot
 Site path              :  sites/default
```

Any thoughts?

Cheers

**Comment by weitzman** [comment]
You have an installed Drupal that works via the web but not in Drush? Does this work: `drush sql-cli`. It should put you in mysql prompt. Also, pass --debug for more verbose logging.

**Comment by welly** [comment]
Yep, works fine via the web. Ran drush status again with --debug:

```
-bash-4.1$ /home/web/example/build/bin/drush --root=/home/web/example/build/docroot/ --debug status
Starting Drush preflight. [0.01 sec, 5.61 MB]                                                                                                                                                                                    [preflight]
Cache MISS cid: 7.0.0-commandfiles-0-6aa6632275ac868143ed7f000e7fe89c [0.01 sec, 5.71 MB]                                                                                                                                            [debug]
Cache SET cid: 7.0.0-commandfiles-0-6aa6632275ac868143ed7f000e7fe89c [0.02 sec, 5.72 MB]                                                                                                                                             [debug]
Bootstrap to phase 0. [0.19 sec, 13.22 MB]                                                                                                                                                                                       [bootstrap]
Drush bootstrap phase : bootstrap_drupal_root() [0.2 sec, 14.02 MB]                                                                                                                                                              [bootstrap]
Initialized Drupal 7.38 root directory at /home/web/example/build/docroot [0.2 sec, 14.02 MB]                                                                                                                                   [notice]
Find command files for phase 1 (max=7) [0.2 sec, 11.68 MB]                                                                                                                                                                           [debug]
Cache MISS cid: 7.0.0-commandfiles-1-867556a8feb2f84c93acab4a36bb2084 [0.2 sec, 11.69 MB]                                                                                                                                            [debug]
Cache SET cid: 7.0.0-commandfiles-1-867556a8feb2f84c93acab4a36bb2084 [0.2 sec, 11.69 MB]                                                                                                                                             [debug]
Drush bootstrap phase : bootstrap_drupal_site() [0.21 sec, 11.69 MB]                                                                                                                                                             [bootstrap]
Initialized Drupal site default at sites/default [0.21 sec, 11.7 MB]                                                                                                                                                                [notice]
Find command files for phase 2 (max=7) [0.21 sec, 11.7 MB]                                                                                                                                                                           [debug]
Cache MISS cid: 7.0.0-commandfiles-2-f17fc6bdab2f74278df75194ca483893 [0.21 sec, 11.7 MB]                                                                                                                                            [debug]
Cache SET cid: 7.0.0-commandfiles-2-f17fc6bdab2f74278df75194ca483893 [0.22 sec, 11.7 MB]                                                                                                                                             [debug]
Drush bootstrap phase : bootstrap_drupal_configuration() [0.22 sec, 11.7 MB]                                                                                                                                                     [bootstrap]
Find command files for phase 3 (max=7) [0.22 sec, 11.71 MB]                                                                                                                                                                          [debug]
Found command: core-status (commandfile=core) [0.22 sec, 11.73 MB]                                                                                                                                                               [bootstrap]
Loading outputformat engine. [0.33 sec, 14.11 MB]                                                                                                                                                                                   [notice]
Calling hook drush_core_status [0.33 sec, 14.16 MB]                                                                                                                                                                                  [debug]
Returned from hook drush_core_status [0.33 sec, 14.18 MB]                                                                                                                                                                            [debug]
 Drupal version         :  7.38
 Site URI               :  http://default
 PHP executable         :  /usr/bin/php
 PHP configuration      :  /etc/php.ini
 PHP OS                 :  Linux
 Drush script           :  /home/web/example/build/vendor/drush/drush/drush.php
 Drush version          :  7.0.0
 Drush temp directory   :  /tmp
 Drush configuration    :
 Drush alias files      :
 Drupal root            :  /home/web/example/build/docroot
 Site path              :  sites/default

Command dispatch complete [0.35 sec, 14.49 MB]                                                                                                                                                                                      [notice]
 Timer  Cum (sec)  Count  Avg (msec)
 page   0.127      1      127.44
```

It's peculiar because it looks like it's going through all the processes correctly from what I can see.

And with drush sql-cli:

```
-bash-4.1$ /home/web/example/build/bin/drush --root=/home/web/example/build/docroot/ --debug sql-cli
Starting Drush preflight. [0.01 sec, 5.61 MB]                                                                                                                                                                                    [preflight]
Cache HIT cid: 7.0.0-commandfiles-0-6aa6632275ac868143ed7f000e7fe89c [0.01 sec, 5.71 MB]                                                                                                                                             [debug]
Bootstrap to phase 0. [0.18 sec, 13.21 MB]                                                                                                                                                                                       [bootstrap]
Bootstrap to phase -1. [0.19 sec, 13.22 MB]                                                                                                                                                                                      [bootstrap]
Found command: sql-cli (commandfile=sql) [0.19 sec, 13.22 MB]                                                                                                                                                                    [bootstrap]
Bootstrap to phase -1. [0.19 sec, 13.22 MB]                                                                                                                                                                                      [bootstrap]
Calling hook drush_sql_cli [0.19 sec, 13.25 MB]                                                                                                                                                                                      [debug]
Drush bootstrap phase : bootstrap_drupal_root() [0.2 sec, 14.05 MB]                                                                                                                                                              [bootstrap]
Initialized Drupal 7.38 root directory at /home/web/example/build/docroot [0.2 sec, 14.05 MB]                                                                                                                                   [notice]
Find command files for phase 1 (max=3) [0.21 sec, 11.68 MB]                                                                                                                                                                          [debug]
Cache HIT cid: 7.0.0-commandfiles-1-867556a8feb2f84c93acab4a36bb2084 [0.21 sec, 11.69 MB]                                                                                                                                            [debug]
Drush bootstrap phase : bootstrap_drupal_site() [0.21 sec, 11.69 MB]                                                                                                                                                             [bootstrap]
Initialized Drupal site default at sites/default [0.21 sec, 11.69 MB]                                                                                                                                                               [notice]
Find command files for phase 2 (max=3) [0.22 sec, 11.69 MB]                                                                                                                                                                          [debug]
Cache MISS cid: 7.0.0-install_profile-66ecfeb9791a023150773849f1550c5d [0.22 sec, 11.7 MB]                                                                                                                                           [debug]
Cache MISS cid: 7.0.0-commandfiles-2-adcbc259623838c285e92ee0febd23c7 [0.22 sec, 11.71 MB]                                                                                                                                           [debug]
Cache SET cid: 7.0.0-commandfiles-2-adcbc259623838c285e92ee0febd23c7 [0.32 sec, 11.72 MB]                                                                                                                                            [debug]
Drush bootstrap phase : bootstrap_drupal_configuration() [0.33 sec, 13.3 MB]                                                                                                                                                     [bootstrap]
Find command files for phase 3 (max=3) [0.33 sec, 13.31 MB]                                                                                                                                                                          [debug]
exception 'ReflectionException' with message 'Class Drush\Sql\Sql7 does not have a constructor, so you cannot pass any constructor arguments' in /home/web/example/build/vendor/drush/drush/includes/drush.inc:113           [error]
Stack trace:
#0 /home/web/example/build/vendor/drush/drush/includes/drush.inc(113): ReflectionClass->newInstanceArgs(Array)
#1 /home/web/example/build/vendor/drush/drush/commands/sql/sql.drush.inc(605): drush_get_class('Drush\Sql\Sql', Array, Array)
#2 /home/web/example/build/vendor/drush/drush/commands/sql/sql.drush.inc(588): drush_sql_get_version()
#3 /home/web/example/build/vendor/drush/drush/commands/sql/sql.drush.inc(454): drush_sql_get_class()
#4 [internal function]: drush_sql_cli()
#5 /home/web/example/build/vendor/drush/drush/includes/command.inc(368): call_user_func_array('drush_sql_cli', Array)
#6 /home/web/example/build/vendor/drush/drush/includes/command.inc(219): _drush_invoke_hooks(Array, Array)
#7 [internal function]: drush_command()
#8 /home/web/example/build/vendor/drush/drush/includes/command.inc(187): call_user_func_array('drush_command', Array)
#9 /home/web/example/build/vendor/drush/drush/lib/Drush/Boot/BaseBoot.php(62): drush_dispatch(Array)
#10 /home/web/example/build/vendor/drush/drush/drush.php(70): Drush\Boot\BaseBoot->bootstrap_and_dispatch()
#11 /home/web/example/build/vendor/drush/drush/drush.php(11): drush_main()
#12 {main} [0.33 sec, 13.35 MB]
Command dispatch complete [0.33 sec, 13.3 MB]                                                                                                                                                                                       [notice]
 Timer  Cum (sec)  Count  Avg (msec)
 page   0.003      1      3.34
```

I'm not sure if I'm missing something obvious here.

**Comment by weitzman** [comment]
That second error suggests that your PHP is too old (/usr/bin/php). See #544. Not sure if upgrading will fix the original problem.

**Comment by welly** [comment]
It seems this server has a global copy of drush installed (6.2.0) and so as a test, I just replaced the version of drush that composer installed with that same version. Problem solved - it all works correctly. I'd like to get this server updated, have been conscious that the php and mysql versions are well out of date.

Cheers, all seems good for now.
